Bluewinds improved API docs

bastelfreak opened the pull request:
This PR adds a general.md file which describes the API and the API design, also a nodes.md with example calls, this will be our new template for API endpoints.

Bluewind committed:
Add nodes.md

Bluewind committed:
Add general.md

Commit:
fix a few typos

bastelfreak commented:
got ack from noctux, waiting for another as this is a big change

taxus13 commented:
@killermoehre I like the idea with the list

there are still open TODOs in general.md, at least i would like to see some keywords there

bastelfreak committed:
fix some typos

bastelfreak commented:
implemented the mentioned stuff from @killermoehre, except 
1. "userable" @Bluewind what did you mean here?
2. "documentation of each affected endpoint"

bastelfreak committed:
fix listing

Bluewind commented:
The first one should be "useable" and "documentation of each affected endpoint" is fine.

Commit:
fix typo (userable -> useable)

bastelfreak commented:
fixed the userable typo. Has anybody thoughts for the TODO parts?

Commit:
add vm.md

Commit:
improve englisch description

Commit:
remove useless error messages

Commit:
switch from mb to byte

Commit:
add note about cgroup

Commit:
add note about UUIDs

Commit:
fix database typo

Commit:
update the nesting

Commit:
fix turner

Commit:
add link to storage.md

Bluewind commented on docs/api/v1/vm.md:
```diff
@@ -0,0 +1,57 @@
+    id: int,
+    virt_method: {
+      id: int,
+      name: string,
```
`vm.node_method.virt_method.name` should not be listed here since `virt_method` will get a dedicated endpoint and then documentation of this value would be duplicated. `.id` should remain because it is referenced in the text in the notes section.

Bluewind commented on docs/api/v1/vm.md:
```diff
@@ -0,0 +1,57 @@
+
+### Notes
+
+A `POST` call will create a VM and a network interface without a VLAN, but with a free IPv4 and IPv6 address. You might want to create a blockdevice [(/storage](storage.md)) and assign it to the new VM. The `cputime_limit` refers to the [Qemu driver for cgroups cpu controller](https://libvirt.org/cgroups.html). You can set `vm.node_method.id` OR `vm.node_method.virt_method.id`. The first one spec ...
```
The qemu documentation describes a lot of different values. To which one does cputime_limit actually refer? cpu_shares?

This should either be very clearly defined and easy to find in the qemu docs or it should be described in detail (possible values and their meaning).

Bluewind replied:
Should a new vm really automatically get a network interface? It might be better if the vm_interface endpoint has an easy way to do this rather than doing it automatically here. We don't create storage so we shouldn't create the network either.

Bluewind replied:
How much of the nested object is actually used when creating a new vm? Maybe general.md should say that e.g. apart from special values only the first level is used.

bastelfreak replied:
interface vs block device creation:
there are many options for a blockdevice (local, ceph, cache options), also a block device isn't nessecary to use the VM.
A network interface doesn't have these options and it is needed to use the virtual machine

nesting level:
we should display all levels that are necessary to create a new object

cputime_limit: added more infos about that feature

Bluewind commented on docs/api/v1/vm.md:
```diff
@@ -0,0 +1,57 @@
+  customer_id: int,
+  cputime_limit: int,
+  uuid: UUID (String),
+  state: {
```
We should make it clear that this is a different state than the one in node.md

bastelfreak replied:
moved it to vm_state and node_state

bastelfreak committed:
update state nesting

bastelfreak committed:
add more infos about libvirt limits

bastelfreak committed:
remove legacy docs

bastelfreak committed:
add node_state and vm_state endpoint

bastelfreak committed:
rename two collums

storage_type.storage_type -> storage_type.name
cache_option.cache_type -> cache_option.name

bastelfreak committed:
add docs for storage endpoints

bastelfreak committed:
enable json highlighting

bastelfreak committed:
enable js highlighting for json

bastelfreak committed:
rename two collums (MB/s -> B/s)

read_mbps_limit -> read_bps_limit
write_mbps_limit -> write_bps_limit

bastelfreak committed:
improve description, update highlighting

bastelfreak committed:
add notes about nullable attributes

bastelfreak committed:
extend notes about nullable attributes

bastelfreak committed:
add virt_method endpoint

bastelfreak committed:
add virt_node endpoint

bastelfreak committed:
don't reference to the referencetable

bastelfreak committed:
add example API call

bastelfreak committed:
add example API call

bastelfreak committed:
add missing whitespace

bastelfreak committed:
add example API call

Commit:
add example API call

bastelfreak committed:
test bash highlighting

bastelfreak committed:
fix bash highlighting

bastelfreak commented:
got accepted on IRC
